### DISCUSSION TITLE
Need help understanding results from text_categorizer

### DISCUSSION THREAD

**Discussion by thomasdhuw** [explosion/spaCy#12281] [Help: Model Advice] [answered]
Hi,

I've just started experimenting with text classification and tried the example below which I found in the docs here - [https://spacy.io/universe/project/classyclassification](url)

```
import spacy
import classy_classification

data = {
    "furniture": ["This text is about chairs.",
                  "Couches, benches and televisions.",
                  "I really need to get a new sofa."],
    "kitchen": ["There also exist things like fridges.",
                "I hope to be getting a new stove today.",
                "Do you also have some ovens."]
}

nlp = spacy.load('en_core_web_md')
nlp.add_pipe("text_categorizer",
             config={
                 "data": data,
                 "model": "spacy"
             }
             )

print(nlp("I am looking for kitchen appliances.")._.cats)
print(nlp("There also exist things like fridges.")._.cats)
```
For the first sentence, the score was much more balanced than the example page. The webpage suggested a score of around .8 for kitchen, but I got the following:

```
{'furniture': 0.45714302933893564, 'kitchen': 0.5428569706610645}
```

For the second nlp call, I expected a very high probability that the category is kitchen given that the string is one of the kitchen example one, but it chose furniture as the highest score (score below).

```
{'furniture': 0.7059043495709844, 'kitchen': 0.29409565042901564}
```

Am I doing something wrong?
I'm using spacy 3.5, class-classification 0.6.1

Thanks

**Comment by kadarakos** [comment, marked as answer]
Hey,

I am not one of the authors/maintainers of `classy-classification`so I cannot give a very in-depth analysis. First, let me say to me it seems like doing everything correctly. 

### The gist of classy

In my understanding the first step in the `classy-classification` workflow when combined with `spaCy` is to extract the document vector produced by the `spacy` pipeline: https://github.com/Pandora-Intelligence/classy-classification/blob/main/classy_classification/classifiers/classy_spacy.py#L97. 

Then it uses an `sklearn` based pipeline with a support-vector machine and grid-search to train a classifier on top of the document representations: https://github.com/Pandora-Intelligence/classy-classification/blob/8aaf887581a8b4f2ec1e2a8bac0b7c8284af262b/classy_classification/classifiers/classy_skeleton.py#L140

### Reproducing your results

I installed the same version of `spacy` and `classy-classification` and reproduced your results, but saw some different numbers. The next step was to modify your code to get a better picture of the non-determinism of the approach and implementation. The modification is just to run your code 10 times and see the results:

```python
import spacy
import classy_classification

data = {
    "furniture": ["This text is about chairs.",
                  "Couches, benches and televisions.",
                  "I really need to get a new sofa."],
    "kitchen": ["There also exist things like fridges.",
                "I hope to be getting a new stove today.",
                "Do you also have some ovens."]
}

for i in range(10):
    nlp = spacy.load('en_core_web_md')
    nlp.add_pipe("text_categorizer",
                 config={
                     "data": data,
                     "model": "spacy"
                 }
                 )

    print(f"Trial {i}")
    print(nlp("I am looking for kitchen appliances.")._.cats)
    print(nlp("There also exist things like fridges.")._.cats)
```

The output generated by the code is this:

```
Trial 0
{'furniture': 0.4826670088770191, 'kitchen': 0.5173329911229809}
{'furniture': 0.8856756028625399, 'kitchen': 0.1143243971374601}
Trial 1
{'furniture': 0.4629880156484637, 'kitchen': 0.5370119843515363}
{'furniture': 0.8641998403871703, 'kitchen': 0.13580015961282943}
Trial 2
{'furniture': 0.44209595031808485, 'kitchen': 0.557904049681915}
{'furniture': 0.8510151415141258, 'kitchen': 0.148984858485874}
Trial 3
{'furniture': 0.46078031139401493, 'kitchen': 0.539219688605985}
{'furniture': 0.8326982650980128, 'kitchen': 0.16730173490198738}
Trial 4
{'furniture': 0.2719839167287487, 'kitchen': 0.7280160832712513}
{'furniture': 0.9952561373295631, 'kitchen': 0.004743862670436858}
Trial 5
{'furniture': 0.4826670088770191, 'kitchen': 0.5173329911229809}
{'furniture': 0.8856756028625399, 'kitchen': 0.1143243971374601}
Trial 6
{'furniture': 0.4522874455390844, 'kitchen': 0.5477125544609155}
{'furniture': 0.8577987830569787, 'kitchen': 0.1422012169430212}
Trial 7
{'furniture': 0.2719839167287487, 'kitchen': 0.7280160832712513}
{'furniture': 0.9952561373295631, 'kitchen': 0.004743862670436858}
Trial 8
{'furniture': 0.4608141153170531, 'kitchen': 0.539185884682947}
{'furniture': 0.8307034413227403, 'kitchen': 0.16929655867725935}
Trial 9
{'furniture': 0.5, 'kitchen': 0.5}
{'furniture': 0.8457007291115707, 'kitchen': 0.15429927088842924}
```

For most of the rounds I got the same `0.5`ish score you've got for `"kitchen"`, but I got lucky with `"Trial 7"` and got the more intuitive `.72`. That is quite a large range! 

Unfortunately also in my trials the kitchen never got a high score: as you can see it never even exceeded `0.17`. 

### Conclusion

It seems like non-determinism is a big factor in the results with `classy-classification`, which makes sense unfortunately with few-shot learning: very few examples and very many variables. It didn't seem to me that you did anything incorrectly. I would recommend trying out `classy-classification` in a more realistic setting with more examples and check how it works and if it doesn't seem to work for you then raising the issue with the authors: https://github.com/Pandora-Intelligence/classy-classification.

**Comment by thomasdhuw** [comment]
Thanks for the feedback, I'll play around with bigger samples as you suggest. I was expecting to see similar numbers to the sample I was following (the .8 for kitchen), but maybe that's not what I should expect.
